ext/http/response_body.rs

```diff
@@ -0,0 +1,247 @@
+  V8Stream(tokio::sync::mpsc::Receiver<BufView>),
+}
+
+impl std::fmt::Debug for ResponseBytesInner {
+  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+    match self {
+      Self::Done => f.write_str("Done"),
+      Self::Empty => f.write_str("Empty"),
+      Self::Bytes(..) => f.write_str("Bytes"),
+      Self::Resource(..) => f.write_str("Resource"),
+      Self::V8Stream(..) => f.write_str("V8Stream"),
+    }
+  }
+}
+
+/// This represents the union of possible response types in Deno with the stream-style [`Body`] interface
+/// required by hyper. As the API requires information about request completion (including a success/fail
+/// flag), we include a very lightweight [`CompletionHandle`] for interested parties to listen on.
+#[derive(Debug, Default)]
+pub struct ResponseBytes(ResponseBytesInner, CompletionHandle);
+
+impl ResponseBytes {
+  pub fn initialize(&mut self, inner: ResponseBytesInner) {
+    debug_assert!(matches!(self.0, ResponseBytesInner::Empty));
+    self.0 = inner;
+  }
+
+  pub fn completion_handle(&self) -> CompletionHandle {
+    self.1.clone()
+  }
+
+  fn complete(&mut self, success: bool) -> ResponseBytesInner {
+    if matches!(self.0, ResponseBytesInner::Done) {
+      return ResponseBytesInner::Done;
+    }
+
+    let current = std::mem::replace(&mut self.0, ResponseBytesInner::Done);
+    self.1.complete(success);
+    current
+  }
+}
+
+impl ResponseBytesInner {
+  pub fn len(&self) -> Option<usize> {
+    match self {
+      Self::Done => Some(0),
+      Self::Empty => Some(0),
+      Self::Bytes(bytes) => Some(bytes.len()),
+      Self::Resource(..) => None,
+      Self::V8Stream(..) => None,
+    }
+  }
+}
+
+impl Body for ResponseBytes {
+  type Data = BufView;
+  type Error = AnyError;
+
+  fn poll_frame(
+    mut self: Pin<&mut Self>,
+    cx: &mut std::task::Context<'_>,
+  ) -> std::task::Poll<Option<Result<Frame<Self::Data>, Self::Error>>> {
+    match &mut self.0 {
+      ResponseBytesInner::Done | ResponseBytesInner::Empty => {
+        unreachable!()
+      }
+      ResponseBytesInner::Bytes(..) => {
+        if let ResponseBytesInner::Bytes(data) = self.complete(true) {
+          std::task::Poll::Ready(Some(Ok(Frame::data(data))))
+        } else {
+          unreachable!()
+        }
+      }
+      ResponseBytesInner::Resource(auto_close, stm, ref mut future) => {
+        match future.poll_unpin(cx) {
+          std::task::Poll::Pending => std::task::Poll::Pending,
+          std::task::Poll::Ready(Err(err)) => {
+            std::task::Poll::Ready(Some(Err(err)))
+          }
+          std::task::Poll::Ready(Ok(buf)) => {
+            if buf.is_empty() {
+              if *auto_close {
+                stm.clone().close();
+              }
+              self.complete(true);
+              return std::task::Poll::Ready(None);
+            }
+            // Re-arm the future
+            *future = stm.clone().read(64 * 1024);
```

Comment:
I am not convinced always using 64kb is good.

deno_core's `op_read_all` dynamically grows the buffer over time until diminishing returns, and uses `resource.size_hint` to inform an optimal starting buffer size. Something like this would probably be useful here too.

https://github.com/denoland/deno/blob/main/core/ops_builtin.rs#L209-L275

Reply:
64kB is exactly what's being read/written in `op_http_write_resource` right now so we're not really losing anything there, and there's already quite a bit of work in this PR. 

It's a big mess interacting with streams that live in v8 right now and I'd prefer an approach where we can wrap v8-side streams in a resource and then use a common method to pump a pipe on the Rust side instead.

Reply:
Yeah, totally. I'll keep this for a future task as we might see some wins here for some tasks.